tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -300,7 +380,7 @@ public int numElements() {
   }
 
   /**
-   * Returns the value in a scalar {@link DataType#FLOAT} tensor.
+   * Returns the value in a scalar {@link TFFloat} tensor.
```

Comment:
Perhaps leave these comments (and the ones above) unchanged?
I'm thinking that the type-system of a Tensor should always refer to the `DataType` (i.e., the return value of `Tensor.dataType()`). So, this does return the value in a scalar `DataType#FLOAT` tensor, and something like `Tensor<Float>` or `Tensor<TFFloat>` has a "type" of `DataType.FLOAT`.

Does that sound sensible? Separating the "type of the tensor" (from the TensorFlow type system) from the "type parameter" of the class.

Reply:
With the design I've proposed, the type DataType is actually mostly invisible and programmers can just use class objects to represent the type of tensors at run time -- rather than having two different notions of type floating around the system. That seems like a win to me.

Now, you might think that we should keep DataType as a very visible entity in the API because of existing code, but as far as I can tell, existing code tends not to talk about DataType because there is no need to: the existing creators in Tensor already automatically decide on datatypes using the object provided at construction time. It seems to me that the stratification you propose will actually make DataType *more* prominent than it currently is, resulting in programmers having two separate ways to think about type.